{
 "metadata": {
  "name": "",
  "signature": "sha256:9eaee44d17ca2f02dcf215a188c911bfe4b532b10e39ee08ad487f625facdf76"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gtf = 'transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"'\n",
      "gtf_list = gtf.split(\" \")\n",
      "print gtf_list\n",
      "print gtf_list[8].find('ENS')\n",
      "print gtf_list[8][1:].find('\"')\n",
      "def get_ensemble_id(gtf_list):\n",
      " \n",
      "    if len(gtf_list) == 9:\n",
      " \n",
      "        ##ids should all start with ENS\n",
      "        ensID_start = gtf_list[8].find('ENS')\n",
      "        ##the id ends with a \", so find the next \" after the start\n",
      "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
      "        return gtf_list[8][ensID_start:ensID_end]\n",
      "    \n",
      "get_ensemble_id(gtf_list)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "['transcribed_unprocessed_pseudogene', 'gene', '11869', '14409', '.', '+', '.', 'gene_id', '\"ENSG00000223972\"']\n",
        "1\n",
        "15\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 19,
       "text": [
        "'ENSG00000223972'"
       ]
      }
     ],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "##input: a list of strings gtf_list (from a line of a GTF file)\n",
      "##output: the ensemble id from gtf_list\n",
      "def get_ensemble_id(gtf_list):\n",
      " \n",
      "    if is_valid_gtf(gtf_list):\n",
      "        return find_ensID(gtf_list)\n",
      " \n",
      "##input: a list of strings gtf_list (from a line of a GTF file)\n",
      "##output: a boolean, True if the list has 9 elements\n",
      "def is_valid_gtf(gtf_list):\n",
      "    return len(gtf_list) == 9\n",
      "    ## GTF files have 9 fields\n",
      "    ##http://uswest.ensembl.org/info/website/upload/gff.html\n",
      " \n",
      "##input: a list of strings gtf_list (from a line of a GTF file)\n",
      "##output: the ensemble id from gtf_list\n",
      "def find_ensID(gtf_list):\n",
      " \n",
      "    annotation = gtf_list[8]\n",
      "    ##ids should all start with ENS\n",
      "    ensID_start = annotation.find('ENS')\n",
      "    ##the id ends with a \", so find the next \" after the start\n",
      "    ensID_end = annotation[ensID_start:].find('\"') + ensID_start\n",
      "    return annotation[ensID_start:ensID_end]\n",
      "\n",
      "print get_ensemble_id(gtf_list)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "ENSG00000223972\n"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "help(dir)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Help on built-in function dir in module __builtin__:\n",
        "\n",
        "dir(...)\n",
        "    dir([object]) -> list of strings\n",
        "    \n",
        "    If called without an argument, return the names in the current scope.\n",
        "    Else, return an alphabetized list of names comprising (some of) the attributes\n",
        "    of the given object, and of attributes reachable from it.\n",
        "    If the object supplies a method named __dir__, it will be used; otherwise\n",
        "    the default dir() logic is used and returns:\n",
        "      for a module object: the module's attributes.\n",
        "      for a class object:  its attributes, and recursively the attributes\n",
        "        of its bases.\n",
        "      for any other object: its attributes, its class's attributes, and\n",
        "        recursively the attributes of its class's base classes.\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dir(gtf)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 24,
       "text": [
        "['__add__',\n",
        " '__class__',\n",
        " '__contains__',\n",
        " '__delattr__',\n",
        " '__doc__',\n",
        " '__eq__',\n",
        " '__format__',\n",
        " '__ge__',\n",
        " '__getattribute__',\n",
        " '__getitem__',\n",
        " '__getnewargs__',\n",
        " '__getslice__',\n",
        " '__gt__',\n",
        " '__hash__',\n",
        " '__init__',\n",
        " '__le__',\n",
        " '__len__',\n",
        " '__lt__',\n",
        " '__mod__',\n",
        " '__mul__',\n",
        " '__ne__',\n",
        " '__new__',\n",
        " '__reduce__',\n",
        " '__reduce_ex__',\n",
        " '__repr__',\n",
        " '__rmod__',\n",
        " '__rmul__',\n",
        " '__setattr__',\n",
        " '__sizeof__',\n",
        " '__str__',\n",
        " '__subclasshook__',\n",
        " '_formatter_field_name_split',\n",
        " '_formatter_parser',\n",
        " 'capitalize',\n",
        " 'center',\n",
        " 'count',\n",
        " 'decode',\n",
        " 'encode',\n",
        " 'endswith',\n",
        " 'expandtabs',\n",
        " 'find',\n",
        " 'format',\n",
        " 'index',\n",
        " 'isalnum',\n",
        " 'isalpha',\n",
        " 'isdigit',\n",
        " 'islower',\n",
        " 'isspace',\n",
        " 'istitle',\n",
        " 'isupper',\n",
        " 'join',\n",
        " 'ljust',\n",
        " 'lower',\n",
        " 'lstrip',\n",
        " 'partition',\n",
        " 'replace',\n",
        " 'rfind',\n",
        " 'rindex',\n",
        " 'rjust',\n",
        " 'rpartition',\n",
        " 'rsplit',\n",
        " 'rstrip',\n",
        " 'split',\n",
        " 'splitlines',\n",
        " 'startswith',\n",
        " 'strip',\n",
        " 'swapcase',\n",
        " 'title',\n",
        " 'translate',\n",
        " 'upper',\n",
        " 'zfill']"
       ]
      }
     ],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash\n",
      "ls -l"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "total 64\n",
        "-rw-r--r--  1 DONG  staff  11599 Jul 24 23:22 7_functions_modules_and_code_reuse.ipynb\n",
        "-rw-r--r--  1 DONG  staff   6665 Jul 25 09:54 8_Interactive Python_testing_and_more_useful_modules.ipynb\n",
        "-rwxr-xr-x@ 1 DONG  staff    280 Jul 24 22:52 greeting_module.py\n",
        "-rw-r--r--  1 DONG  staff    549 Jul 24 22:53 greeting_module.pyc\n",
        "-rwxr-xr-x@ 1 DONG  staff     74 Jul 24 08:12 testprogram.py\n"
       ]
      }
     ],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%pwd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 33,
       "text": [
        "u'/Users/DONG/Documents/Python/Python_world/4_reading_writing_using_lots_of_data'"
       ]
      }
     ],
     "prompt_number": 33
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%hist -nt"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "   1:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf\n",
        "   2:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "   3:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   4:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   5:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   6:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   7:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   8:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][ensID_start:]\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   9:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][1:]\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  10:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][1:9]\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  11:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][1:9].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  12:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  13:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  14:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][0:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  15:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][2:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  16:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  17:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  18:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  19:\n",
        "gtf = 'transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"'\n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  20:\n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: the ensemble id from gtf_list\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if is_valid_gtf(gtf_list):\n",
        "        return find_ensID(gtf_list)\n",
        " \n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: a boolean, True if the list has 9 elements\n",
        "def is_valid_gtf(gtf_list):\n",
        "    return len(gtf_list) == 9\n",
        "    ## GTF files have 9 fields\n",
        "    ##http://uswest.ensembl.org/info/website/upload/gff.html\n",
        " \n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: the ensemble id from gtf_list\n",
        "def find_ensID(gtf_list):\n",
        " \n",
        "    annotation = gtf_list[8]\n",
        "    ##ids should all start with ENS\n",
        "    ensID_start = annotation.find('ENS')\n",
        "    ##the id ends with a \", so find the next \" after the start\n",
        "    ensID_end = annotation[ensID_start:].find('\"') + ensID_start\n",
        "    return annotation[ensID_start:ensID_end]\n",
        "  21:\n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: the ensemble id from gtf_list\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if is_valid_gtf(gtf_list):\n",
        "        return find_ensID(gtf_list)\n",
        " \n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: a boolean, True if the list has 9 elements\n",
        "def is_valid_gtf(gtf_list):\n",
        "    return len(gtf_list) == 9\n",
        "    ## GTF files have 9 fields\n",
        "    ##http://uswest.ensembl.org/info/website/upload/gff.html\n",
        " \n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: the ensemble id from gtf_list\n",
        "def find_ensID(gtf_list):\n",
        " \n",
        "    annotation = gtf_list[8]\n",
        "    ##ids should all start with ENS\n",
        "    ensID_start = annotation.find('ENS')\n",
        "    ##the id ends with a \", so find the next \" after the start\n",
        "    ensID_end = annotation[ensID_start:].find('\"') + ensID_start\n",
        "    return annotation[ensID_start:ensID_end]\n",
        "\n",
        "print get_ensemble_id(gtf_list)\n",
        "  22: help(dir)\n",
        "  23: dir(gtf_list)\n",
        "  24: dir(gtf)\n",
        "  25: get_ipython().magic(u'ls ')\n",
        "  26: get_ipython().run_cell_magic(u'bash', u'', u'ls -l')\n",
        "  27: get_ipython().magic(u'history')\n",
        "  28: get_ipython().magic(u'ls')\n",
        "  29: get_ipython().magic(u'la')\n",
        "  30: get_ipython().magic(u'cd ..')\n",
        "  31: get_ipython().magic(u'pwd')\n",
        "  32: get_ipython().magic(u'cd /Users/DONG/Documents/Python/Python_world/4_reading_writing_using_lots_of_data/')\n",
        "  33: get_ipython().magic(u'pwd')\n",
        "  34: get_ipython().magic(u'hist -nt')\n"
       ]
      }
     ],
     "prompt_number": 34
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%save histfile.py 32-34"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "File `histfile.py` exists. Overwrite (y/[N])?  y\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "The following commands were written to file `histfile.py`:\n",
        "get_ipython().magic(u'cd /Users/DONG/Documents/Python/Python_world/4_reading_writing_using_lots_of_data/')\n",
        "get_ipython().magic(u'pwd')\n",
        "get_ipython().magic(u'hist -nt')\n"
       ]
      }
     ],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "run histfile"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/Users/DONG/Documents/Python/Python_world/4_reading_writing_using_lots_of_data\n",
        "   1:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf\n",
        "   2:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "   3:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   4:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   5:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   6:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   7:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   8:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][ensID_start:]\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "   9:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][1:]\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  10:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][1:9]\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  11:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][1:9].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  12:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  13:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  14:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][0:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  15:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][2:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  16:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "#print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  17:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    #if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  18:\n",
        "gtf = '''transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"''' \n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  19:\n",
        "gtf = 'transcribed_unprocessed_pseudogene gene 11869 14409 . + . gene_id \"ENSG00000223972\"'\n",
        "gtf_list = gtf.split(\" \")\n",
        "print gtf_list\n",
        "print gtf_list[8].find('ENS')\n",
        "print gtf_list[8][1:].find('\"')\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if len(gtf_list) == 9:\n",
        " \n",
        "        ##ids should all start with ENS\n",
        "        ensID_start = gtf_list[8].find('ENS')\n",
        "        ##the id ends with a \", so find the next \" after the start\n",
        "        ensID_end = gtf_list[8][ensID_start:].find('\"') + ensID_start\n",
        "        return gtf_list[8][ensID_start:ensID_end]\n",
        "    \n",
        "get_ensemble_id(gtf_list)\n",
        "  20:\n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: the ensemble id from gtf_list\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if is_valid_gtf(gtf_list):\n",
        "        return find_ensID(gtf_list)\n",
        " \n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: a boolean, True if the list has 9 elements\n",
        "def is_valid_gtf(gtf_list):\n",
        "    return len(gtf_list) == 9\n",
        "    ## GTF files have 9 fields\n",
        "    ##http://uswest.ensembl.org/info/website/upload/gff.html\n",
        " \n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: the ensemble id from gtf_list\n",
        "def find_ensID(gtf_list):\n",
        " \n",
        "    annotation = gtf_list[8]\n",
        "    ##ids should all start with ENS\n",
        "    ensID_start = annotation.find('ENS')\n",
        "    ##the id ends with a \", so find the next \" after the start\n",
        "    ensID_end = annotation[ensID_start:].find('\"') + ensID_start\n",
        "    return annotation[ensID_start:ensID_end]\n",
        "  21:\n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: the ensemble id from gtf_list\n",
        "def get_ensemble_id(gtf_list):\n",
        " \n",
        "    if is_valid_gtf(gtf_list):\n",
        "        return find_ensID(gtf_list)\n",
        " \n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: a boolean, True if the list has 9 elements\n",
        "def is_valid_gtf(gtf_list):\n",
        "    return len(gtf_list) == 9\n",
        "    ## GTF files have 9 fields\n",
        "    ##http://uswest.ensembl.org/info/website/upload/gff.html\n",
        " \n",
        "##input: a list of strings gtf_list (from a line of a GTF file)\n",
        "##output: the ensemble id from gtf_list\n",
        "def find_ensID(gtf_list):\n",
        " \n",
        "    annotation = gtf_list[8]\n",
        "    ##ids should all start with ENS\n",
        "    ensID_start = annotation.find('ENS')\n",
        "    ##the id ends with a \", so find the next \" after the start\n",
        "    ensID_end = annotation[ensID_start:].find('\"') + ensID_start\n",
        "    return annotation[ensID_start:ensID_end]\n",
        "\n",
        "print get_ensemble_id(gtf_list)\n",
        "  22: help(dir)\n",
        "  23: dir(gtf_list)\n",
        "  24: dir(gtf)\n",
        "  25: get_ipython().magic(u'ls ')\n",
        "  26: get_ipython().run_cell_magic(u'bash', u'', u'ls -l')\n",
        "  27: get_ipython().magic(u'history')\n",
        "  28: get_ipython().magic(u'ls')\n",
        "  29: get_ipython().magic(u'la')\n",
        "  30: get_ipython().magic(u'cd ..')\n",
        "  31: get_ipython().magic(u'pwd')\n",
        "  32: get_ipython().magic(u'cd /Users/DONG/Documents/Python/Python_world/4_reading_writing_using_lots_of_data/')\n",
        "  33: get_ipython().magic(u'pwd')\n",
        "  34: get_ipython().magic(u'hist -nt')\n",
        "  35: get_ipython().magic(u'save histfile.py 19-24')\n",
        "  36: get_ipython().magic(u'save histfile.py 1')\n",
        "  37: get_ipython().magic(u'run histfile')\n",
        "  38: get_ipython().magic(u'save histfile.py 21-22')\n",
        "  39: get_ipython().magic(u'run histfile')\n",
        "  40: get_ipython().magic(u'run histfile')\n",
        "  41: get_ipython().magic(u'save histfile.py 32-34')\n",
        "  42: get_ipython().magic(u'run histfile')\n",
        "  43: get_ipython().magic(u'ls ')\n",
        "  44: get_ipython().magic(u'run testprogram')\n",
        "  45: get_ipython().magic(u'run testprogram.py')\n",
        "  46: get_ipython().magic(u'save hist.py 32-34')\n",
        "  47: get_ipython().magic(u'run hist')\n",
        "  48: get_ipython().magic(u'edit histfile')\n",
        "  49: get_ipython().magic(u'edit histfile')\n",
        "  50: get_ipython().magic(u'edit histfile.py')\n",
        "  51: get_ipython().magic(u'save histfile.py 32-34')\n",
        "  52: get_ipython().magic(u'run histfile')\n"
       ]
      }
     ],
     "prompt_number": 52
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%edit histfile"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 53
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gtf_list.append?"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 62
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def fastaParser(filename):\n",
      "   current_gene = \"\"\n",
      "   genes = {}\n",
      "   fh = open(filename, 'r')\n",
      " \n",
      "   for line in fh:\n",
      "       line = line.strip()\n",
      "       if line.startswith('>'):\n",
      "           current_gene = line[1:]\n",
      "           genes[current_gene] = ''\n",
      "       else:\n",
      "            join(genes, current_gene, line)\n",
      " \n",
      "   return genes\n",
      " \n",
      " \n",
      "def join(gene_dict, gene, line):\n",
      "    gene_dict[gene] += line\n",
      " \n",
      "\n",
      "c_elegans = fastaParser('ceWS215_tRNA.fa')\n",
      " \n",
      "print \"There are\", len(c_elegans), \"chromosomes in C.elegans\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "There are 660 chromosomes in C.elegans\n"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def fastaParser2(filename):\n",
      "   current_gene = \"\"\n",
      "   genes = {}\n",
      "   fh = open(filename, 'r')\n",
      " \n",
      "   for line in fh:\n",
      "       line = line.strip()\n",
      "       if line.startswith('>'):\n",
      "           current_gene = line[1:]\n",
      "           genes[current_gene] = []\n",
      "       else:\n",
      "            join2(genes, current_gene, line)\n",
      " \n",
      "   for gene in genes:\n",
      "       genes[gene] = ''.join(genes[gene])\n",
      " \n",
      "   return genes\n",
      " \n",
      " \n",
      "def join2(gene_dict, gene, line):\n",
      "    gene_dict[gene].append(line)\n",
      " \n",
      "yeast_genome = fastaParser2('ceWS215_tRNA.fa')\n",
      " \n",
      "print \"There are\", len(yeast_genome), \"chromosomes in yeast\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "run parse_genome.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "There are 660 chromosomes in C.elegans\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "run -p parse_genome.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}